---
description: "Reports empty type parameter lists in type aliases and interfaces."
title: "emptyTypeParameterLists"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="emptyTypeParameterLists" />

TypeScript permits empty type parameter lists like `<>`, but this practice is unnecessary.
Empty type parameter lists can lead to unclear or ambiguous code where the intention is not self-evident.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
type Foo<> = string;
```

```ts
interface Bar<> {}
```

```ts
type Generic<> = { value: number };
```

</TabItem>
<TabItem label="✅ Correct">

```ts
type Foo = string;
```

```ts
interface Bar {}
```

```ts
type Generic<T> = T;
```

```ts
interface Container<T> {
	value: T;
}
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If your codebase intentionally uses empty type parameter lists for specific tooling or code generation purposes, you may disable this rule.

## Further Reading

- [TypeScript Handbook: Generics](https://www.typescriptlang.org/docs/handbook/2/generics.html)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="emptyTypeParameterLists" />
