---
description: "Reports comparisons with NaN, which should use Number.isNaN() instead."
title: "isNaNComparisons"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="isNaNComparisons" />

In JavaScript, `NaN` (Not-a-Number) has a unique property: it is not equal to anything, including itself.
This means that comparisons like `x === NaN` or `x !== NaN` always produce unexpected results - the equality check always returns `false`, and the inequality check always returns `true`.

To reliably check if a value is `NaN`, use `Number.isNaN()` instead.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
if (value === NaN) {
	console.log("value is NaN");
}
```

```ts
const isNotANumber = result !== NaN;
```

```ts
if (value > NaN) {
	console.log("comparison result");
}
```

</TabItem>
<TabItem label="✅ Correct">

```ts
if (Number.isNaN(value)) {
	console.log("value is NaN");
}
```

```ts
const isNotANumber = !Number.isNaN(result);
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

There is no legitimate reason to compare directly with `NaN`.
This rule should always be enabled to prevent subtle bugs.

## Further Reading

- [MDN: NaN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NaN)
- [MDN: Number.isNaN()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isNaN)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="isNaNComparisons" />
