---
description: "Reports regex patterns that use ambiguous or invalid syntax from Annex B."
title: "regexAmbiguousInvalidity"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="regexAmbiguousInvalidity" />

ECMAScript Annex B defines legacy regex syntax that browsers must support for web compatibility but is considered ambiguous or deprecated.
This includes octal escapes in patterns, incomplete escape sequences, useless escapes, and unescaped special characters outside of character classes.
Using strict regex syntax ensures better clarity and cross-platform compatibility.

When the `u` (unicode) or `v` (unicode sets) flag is present, JavaScript already enforces strict regex parsing, so this rule skips those patterns.

## Examples

<Tabs>
<TabItem label="❌ Incorrect">

```ts
// Octal escapes in regex patterns
const re1 = /\1/;
const re2 = /\07/;

// Incomplete escape sequences
const re3 = /\x1/;
const re4 = /\u123/;
const re5 = /\c1/;

// Unescaped brackets outside character classes
const re6 = /a]/;
const re7 = /a{/;

// Useless escapes
const re8 = /\q/;
const re9 = /\!/;

// Same issues with RegExp constructor
const re10 = new RegExp("\\1");
```

</TabItem>
<TabItem label="✅ Correct">

```ts
// Use hexadecimal escapes instead of octal
const re1 = /\x01/;
const re2 = /\x07/;

// Complete escape sequences
const re3 = /\x1F/;
const re4 = /\u0123/;
const re5 = /\cA/;

// Escape brackets outside character classes
const re6 = /a\]/;
const re7 = /a\{/;

// No useless escapes needed
const re8 = /q/;
const re9 = /!/;

// Unicode mode enforces strict parsing
const re10 = /\1/u;
const re11 = new RegExp("\\1", "u");
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If you are intentionally using legacy regex syntax for compatibility with very old JavaScript environments or specific regex behavior, you may choose to disable this rule.
However, in modern development, using strict regex syntax is recommended.

## Further Reading

- [ECMAScript Annex B: Additional ECMAScript Features for Web Browsers](https://tc39.es/ecma262/#sec-additional-ecmascript-features-for-web-browsers)
- [MDN: Regular expressions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions)
- [MDN: Unicode flag](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/unicode)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="regexAmbiguousInvalidity" />
