---
description: "Reports consecutive identical elements in regular expressions that should use quantifiers."
title: "regexRepeatQuantifiers"
topic: "rules"
---

import { Tabs, TabItem } from "@astrojs/starlight/components";
import { RuleEquivalents } from "~/components/RuleEquivalents";
import RuleSummary from "~/components/RuleSummary.astro";

<RuleSummary plugin="ts" rule="regexRepeatQuantifiers" />

Matching multiple consecutive characters in a regular expression can be done either by writing out that character repeatedly or using a quantifier.
Quantifiers keep the regex shorter and harder to miscount, and they make it obvious how many times a part repeats without visually scanning a long run of characters.
For example, writing `a{5}` is more concise and readable than `aaaaa`.

This rule reports runs of five or more identical elements in regular expressions and suggests replacing them with quantifiers.

## Examples

### Repeated Characters

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /aaaaa/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /a{5}/;
```

</TabItem>
</Tabs>

### Escape Sequences

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /\d\d\d\d\d/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /\d{5}/;
```

</TabItem>
</Tabs>

### Character Classes

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /[a-z][a-z][a-z][a-z][a-z]/;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /[a-z]{5}/;
```

</TabItem>
</Tabs>

### Dots

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = /...../;
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = /.{5}/;
```

</TabItem>
</Tabs>

### RegExp Constructor

<Tabs>
<TabItem label="❌ Incorrect">

```ts
const pattern = new RegExp("aaaaa");
```

</TabItem>
<TabItem label="✅ Correct">

```ts
const pattern = new RegExp("a{5}");
```

</TabItem>
</Tabs>

## Options

This rule is not configurable.

## When Not To Use It

If your project has a style guide that prefers explicit repetition for clarity, or if you are programmatically generating regular expressions where repetition is easier to produce, you might want to disable this rule.

## Further Reading

- [MDN: Quantifier](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions/Quantifiers)
- [MDN: Regular expressions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions)

## Equivalents in Other Linters

<RuleEquivalents pluginId="ts" ruleId="regexRepeatQuantifiers" />
